packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,664 @@
+    if (method in this._permissionSpecifications) {
+      return method;
+    }
+
+    const wildCardMethodsWithoutWildCard: Record<string, boolean> = {};
+    for (const targetKey of Object.keys(this._permissionSpecifications)) {
+      const wildCardMatch = targetKey.match(/(.+)\*$/u);
+      if (wildCardMatch) {
+        wildCardMethodsWithoutWildCard[wildCardMatch[1]] = true;
+      }
+    }
+
+    // Check for potentially nested namespaces:
+    // Ex: wildzone_
+    // Ex: eth_plugin_
+    const segments = method.split('_');
+    let targetKey = '';
+
+    while (
+      segments.length > 0 &&
+      !(targetKey in this._permissionSpecifications) &&
+      !wildCardMethodsWithoutWildCard[targetKey]
+    ) {
+      targetKey += `${segments.shift()}_`;
+    }
+
+    if (targetKey in this._permissionSpecifications) {
+      return targetKey;
+    } else if (wildCardMethodsWithoutWildCard[targetKey]) {
+      return `${targetKey}*`;
+    }
+
+    return '';
+  }
+
+  getRestrictedMethodImplementation(
+    method: string,
+  ): RestrictedMethodImplementation<Json, Json> | undefined {
+    const targetKey = this.getTargetKey(method);
+    if (!targetKey) {
+      return undefined;
+    }
+
+    return this.getPermissionSpecification(method).methodImplementation;
+  }
+
+  private getPermissionSpecification(targetKey: string) {
+    const specification = this._permissionSpecifications[targetKey];
+    /* istanbul ignore if: This should be impossible, but TypeScript wants it */
+    if (!specification) {
+      throw new Error(
+        `Existing permission target key "${targetKey}" has no specification.`,
+      );
+    }
+
+    return specification;
+  }
+
+  grantPermissions(opts: GrantPermissionsOptions): SubjectPermissions {
+    const {
+      requestData,
+      approvedPermissions,
+      shouldPreserveExistingPermissions = false,
+      subject,
+    } = opts;
+    const { origin } = subject;
+
+    if (!origin || typeof origin !== 'string') {
+      throw new InvalidSubjectIdentifierError(origin);
+    }
+
+    const permissions: { [methodName: string]: Permission } =
+      (shouldPreserveExistingPermissions && {
+        ...this.getPermissions(origin),
+      }) ||
+      {};
+
+    for (const target of Object.keys(approvedPermissions)) {
+      const targetKey = this.getTargetKey(target);
+      if (!targetKey) {
+        throw methodNotFound({ method: target });
+      }
+
+      const specification = this.getPermissionSpecification(targetKey);
+
+      const caveats = this.computeCaveats(
+        origin,
+        target,
+        approvedPermissions[target].caveats,
+      );
+
+      const permissionOptions = { caveats, invoker: origin, target };
+
+      let permission: Permission;
+      if (specification.factory) {
+        permission = specification.factory(permissionOptions, requestData);
+      } else {
+        permission = constructPermission(permissionOptions);
+        specification.validator?.(permission);
```

Comment:
I don't follow why we need to validate our own code at runtime :thinking: What is this for? I would expect us to validate untrusted input, but this doesn't appear to be untrusted input. The permission being validated here is what we constructed.

The same goes for the caveats as well.